### ISSUE TITLE
typeclass resolution does not backtrack enough unless Hint Mode + is set

### ISSUE THREAD

**Issue by andres-erbsen** [coq/coq#14125]
#### Description of the problem

I thought adding `Hint Mode +` would solve no more goals than `Hint Mode -`, but here we go:

```coq
Class A (x : True) := a : True.
Class B x (y : A x) := b : True.
Axiom pf : forall x y, B x y -> False.
Instance: B I I := I.
Goal False.
  Hint Mode A + : typeclass_instances.
  pose (pf _ _ _).

  Hint Mode A - : typeclass_instances.
  pose (pf _ _ _).
  (* 
Cannot infer this placeholder of type "B ?x ?y" (no type class instance
found) in environment:
f := pf I I B_instance_0 : False
*)
Abort.
```

Minimized by @JasonGross from https://github.com/mit-plv/riscv-coq/commit/2fd492cf4daca76d2bb96c26a7539e0df67e247e#diff-5ef0ed939b63ace5108e14a7d2a773a485de80c3900243f485d37518a31759baR165

The documentation for `Hint Mode` does not say anything about backtracking.

> The documentation of `typeclasses eauto` says "Unlike eauto and auto, the resolution is done entirely in the proof engine, meaning that backtracking is available among dependent subgoals, and shelving goals is supported. typeclasses eauto is a multi-goal tactic. It analyses the dependencies between subgoals to avoid backtracking on subgoals that are entirely independent."

We can see that the goals here are not entirely independent because the solving order used when `Hint Mode +` is causes both goals to be solved.

#### Coq Version

`The Coq Proof Assistant, version 8.13.1 (March 2021)
compiled on Mar 4 2021 14:01:37 with OCaml 4.11.1`

@mattam82 ?

**Comment by mattam82** [comment]
This is corrected in coq/coq#13952. If we get a "stuck failure" for A because it matches the mode declaration but is unsolvable, we still try to solve the other goals.

**Comment by mattam82** [comment]
The doc also explains how resolution goes in case of mode declarations (the previous implementation indeed did not do enough backtracking)
